README.md
# BAMBOOHR-CONNECTOR

## Introduction

This repository contains the code for the project `BambooHR-connector`, which automatically imports the absences for all employees from BambooHR to Papershift.
## Key Facts

Product Owner / Project Manager: [Christoph Batik](mailto:christoph.batik@bestit-online.at)

Developer: [Georgi Damyanov](mailto:georgi.damyanov@bestit-online.at)

## Development Setup

### Requirements

- MAMP or XAMPP
- PHP 7.1

### Download

- `Git clone` this repository into `MAMP/XAMPP`
- Run `composer install`

### Configuration

To use the connector first you have to configure it.

- Open Config.php
- Fill the missing `keys`.

- `bamboohr`
    - `api_token` - The API key for BambooHR
    - `company` - company name
    - `filter_enabled` - `enables/disables` the absences filter (look `filter_array`)
    - `filter_array` - which absences to be `filtered/removed`
    - `date` - Only the absences from this date will be taken. You can choose between `yesterday`, `today`, `tomorrow` or give `custom Y-m-d`.
    - `status` - Which absences to get. You can choose from `approved`, `denied`, `superceded`, `requested`, `canceled`

- `papershift`
    - `api_token` - The API key for Papershift
    - `api_absences_url` - The Papershift API absences URL
    - `api_absences_confirm_url` - The Papershift API absences/confirm URL
    
    Papershift requires `start` and `end time` for the absences
    
    - `start_time_days` - Start time for the absence
    - `end_time_days` - End time for the absence
    - `full_day` - Full day
    
    Papershift does not allow to enter the absence hours. That is why every time there are  hours (2,3,4 hours) they are counted from `00:00:00` onward. ( the hours are calculated inside `PapershiftAbsences.php`)
    
    - `start_time_hours` - The start time for the hours 
    
- `mail`
    - `smtp_host` - Your e-mail host
    - `smtp_auth` - If Authentication should be enabled
    - `username` - Your e-mail
    - `password` - Your e-mail password
    - `smtp_secure` - Security
    - `port` - Port
    - `send_from` - Sender (e-mail)
    - `send_to` - Recipient (e-mail)

### Run Script

Once you are done with the configuration, to run the script:

- Open `Connector.php` inside your browser
- It will take couple of minutes, which highly depends on how many absences there are
- You will be notified per e-mail if the import was successful or failed

## Testing

There are no tests at this point, you have to test manually.

## Additional Documentation

### Error Logs

There are two logs inside the `log` folder.

- `tmp_error_log.txt` which is used to temporarily store the errors.
- `global_error_log.txt` contains all the errors.

Errors from `tmp_error_log.txt` are send via e-mail. Then it appends the errors to the `global_error_log.txt` and clears itself. Next time the script runs `tmp_error_log` will be empty. That way you don't get all errors per e-mail, but only the errors from the current day.

changelog.txt
- Version 3.0
    - All absences are now being imported from the day that the Cronjob has run forward.

- Version 2.1
    - The connector automatically "confirms" the absences after they are inserted into Papershift. (no need to do it manually anymore)
    - Added .gitignore
    - Updated README.md

- Version 2.0
    - Rewritten from functional to object oriented.

Tests/examples.php
<?php

include "BambooHR/API/API.php";
use \BambooHR\API\BambooAPI as BHR;


//this can be either your subdomain, or the full domain, so either "test", or "test.bamboohr.com"
$api = new BHR('bestitonline');
//Users can create an api key in bamboohr, see http://www.bamboohr.com/api/documentation/#authentication for more details.
$api->setSecretKey('9e81788bfc9cd5901af0e1370d32ab54ad516dc5');
//or if you have contacted BambooHR to get an application key, you can allow users to login with their usernames and passwords. You should not store usernames and passwords, but store keys instead.
$authenticated = $api->login('<your application key>', '<username>', '<password>');
if(!$authenticated) {
	trigger_error("Login Failure");
}

//It's a good idea to take a look at the fields that are available in your account
$response = $api->getFields();
//$response is an object representing the HTTP response. Errors can be detected by testing the statusCode:
if($response->statusCode < 200 || $response->statusCode > 299) {
	trigger_error("Login Failure");
}
//or by using $response->isError()
if($response->isError()) {
	trigger_error("Login Failure");
}
//APIs return in XML format by default. The getContent() function will return a
//SimpleXMLElement object of the root XML element in the response. If JSON is
//returned, it will return a stdclass object.
$xml = $response->getContent();
foreach($xml->field as $field ){
	echo "Field: " . $field;
}

//tables can either be a list of items (like contacts), or a historical record (like job history), but can be accessed in similar ways.
$response = $api->getTables();
//This will allow you to view valid values for "list" fields.
$response = $api->getLists();

$fields = array('lastName', 'firstName', 'employeeNumber');
$response = $api->getCustomReport('xml',$fields);
if($response->isError()){
	//for many errors, Bamboo will add a header "X-BambooHR-Error-Messsage" that describes the error. These messages are meant for debugging purposes.
	trigger_error("Problem with custom report: " . $response->getErrorMessage());
}



Tests/addAbsenceToPapershift.php
<?php
/**
 * Created by PhpStorm.
 * User: gdamyanov
 * Date: 25.10.17
 * Time: 18:01
 */


$data = array(  "api_token" => "dnad0PIgOFixZae0AEnVUl6YwsrpJa1hUYUHcWvD",
                    "absence" => array("absence_type_external_id" => "Urlaub", "user_external_id" =>"79", "starts_at" => "2017-10-24T00:00:00+02:00", "ends_at" => "2017-10-24T24:00:00+02:00", "full_day" => false)
);
$data_string = json_encode($data);

//echo date("c");exit;

$ch = curl_init('https://app.papershift.com/public_api/v1/absences');
curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
curl_setopt($ch, CURLOPT_POSTFIELDS, $data_string);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_HTTPHEADER, array(
        'Content-Type: application/json',
        'Content-Length: ' . strlen($data_string),
        'Accept: application/json')
);
curl_setopt($ch, CURLOPT_TIMEOUT, 5);
curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);

//execute post
$result = curl_exec($ch);

//close connection
curl_close($ch);

echo $result;

Config.php
<?php

declare(strict_types=1);

namespace BambooHRConnector;

/**
 * Class Config
 *
 * @package BambooHRConnector
 * @author Georgi Damyanov <georgi.damyanov@bestit-online.at>
 */
class Config
{
    /**
     * Config array
     *
     * @var $config array
     */
    private $config;

    /**
     * Logger array
     *
     * @var $logger Logger
     */
    private $logger;

    /**
     * Config constructor.
     *
     * @param Logger $logger Logger Instance
     */
    public function __construct(Logger $logger)
    {
        $this->logger = $logger;

        $this->config = [
            'bamboohr' => [
                'api_token' => '8f912006d3d165af282ea16be488d668d2d41915',
                'company' => 'bestitonline',
                'filter_enabled' => true,
                'filter_array' => ['Abwesend ','Abwesend', 'Geschäftsreise', 'Homeoffice'],
                'date' => date('Y-m-d', strtotime('today')),
                'status' => 'approved'
            ],
            'papershift' => [
                'api_token' => '8f912006d3d165af282ea16be488d668d2d41915',
                'api_absences_url' => 'https://app.papershift.com/public_api/v1/absences',
                'api_absences_confirm_url' => 'https://app.papershift.com/public_api/v1/absences/confirm',
                'start_time_days' => 'T08:00:00+01:00',
                'end_time_days' => 'T17:00:00+01:00',
                'start_time_hours' => 'T00:00:00+01:00',
                'full_day' => true,
            ],
            'mail' => [
                'smtp_host' => 'dsadasdsa',
                'smtp_auth' => true,
                'username' => 'dasdsadsa',
                'password' => 'dsadsadsa',
                'smtp_secure' => 'dasdsadsa',
                'port' => 'dsadasdas',
                'send_from' => 'dsadsadsa',
                'send_to' => 'dsadsadsa'
            ]
        ];
    }

    /**
     * Function getConfig
     *
     * Get the config array.
     *
     * @param String $category Name of the category (bamboohr/papershift/mail)
     * @param String $key Name of the key
     *
     * @return Mixed mixed If there is no error it returns Array
     */
    public function getConfig($category, $key)
    {
        if (!array_key_exists($category, $this->config)) {
            $this->logger->writeErrorToFile("The category '".$category."' does not exist!");
            die("The category '".$category."' does not exist!");
        }

        if (!array_key_exists($key, $this->config[$category])) {
            $this->logger->writeErrorToFile("The key '".$key."' inside category '".$category."' does not exist!");
            die("The key '".$key."' inside category '".$category."' does not exist!");
        } else {
            if ($this->config[$category][$key] === '' || is_null($this->config[$category][$key])) {
                $this->logger->writeErrorToFile("The key '".$key."' does not contain any value!");
                die("The key '".$key."' does not contain any value!");
            }
        }
        return $this->config[$category][$key];
    }
}


Logger.php
<?php

declare(strict_types=1);

namespace BambooHRConnector;

use DateTime;
use DateTimeZone;
use Exception;

/**
 * Class Logger
 *
 * @package BambooHRConnector
 * @author Georgi Damyanov <georgi.damyanov@bestit-online.at>
 */
class Logger
{
    /**
     * TMP file containing errors of the day
     *
     * @var $tmp_error_log string
     */
    private $tmp_error_log;

    /**
     * The global error log file, which contains all the errors
     *
     * @var $global_error_log string
     */
    private $global_error_log;

    /**
     * Logger constructor.
     */
    public function __construct()
    {
        $this->tmp_error_log = './log/tmp_error.txt';
        $this->global_error_log = './log/global_error.txt';
    }

    /**
     * Function writeErrorToFile
     *
     * - Get current date
     * - Write the error message in the tmp_error_log file
     *
     * @param string $message The message which will be save in the log file
     *
     * @return bool Returns true if data successfully written
     * @throws Exception
     */
    public function writeErrorToFile($message)
    {
        static $linebreak = false;

        $date = new DateTime();
        $date->setTimezone(new DateTimeZone('Europe/Berlin'));
        $date = $date->format('Y-m-d H:i:sP');

        if (!$linebreak) {
            $lineSeparator = <<< EOT
--------------------------------------------------------------------------------------------------------------------\n
EOT;
            error_log($lineSeparator, 3, $this->tmp_error_log);
            $linebreak = true;
        }
        error_log($date.' - '.$message."\n", 3, $this->tmp_error_log);
        return true;
    }

    /**
     * Function transferDailyLogToGlobal
     *
     * The tmp_error_log is appended to the globa_error.txt and cleared for the next day
     *
     * @return void
     */
    public function transferDailyLogToGlobal(): void
    {
        $this->createLogDirectory();
        $this->createLogFiles();

        if (file_exists($this->tmp_error_log) && filesize($this->tmp_error_log) !== 0) {
            file_put_contents($this->global_error_log, file_get_contents($this->tmp_error_log), FILE_APPEND | LOCK_EX);
            file_put_contents($this->tmp_error_log, '');
        }
    }

    /**
     * Create log directory if it does not exist
     *
     * @return void
     */
    private function createLogDirectory(): void
    {
        if (!file_exists('./log')) {
            mkdir('./log', 0755, true);
        }
    }

    /**
     * Create the global_error and tmp_error files if they don't exist
     *
     * @return void
     */
    private function createLogFiles(): void
    {
        if (!is_file($this->global_error_log)) {
            fopen($this->global_error_log, 'w');
        }

        if (!is_file($this->tmp_error_log)) {
            fopen($this->tmp_error_log, 'w');
        }
    }
}


SendMail.php
<?php

declare(strict_types=1);

namespace BambooHRConnector;

use PHPMailer\PHPMailer\Exception;
use PHPMailer\PHPMailer\PHPMailer;

/**
 * Class SendMail
 *
 * @package BambooHRConnector
 * @author Georgi Damyanov <georgi.damyanov@bestit-online.at>
 */
class SendMail
{
    /**
     * Instance of PHPMailer
     *
     * @var $mail PHPMailer
     */
    private $mail;

    /**
     * Instance of Config
     *
     * @var $config Config
     */
    private $config;

    /**
     * SendMail constructor.
     *
     * @param Config $config Config Instance
     * @param PHPMailer $mail PHPMailer Instance
     */
    public function __construct(Config $config, PHPMailer $mail)
    {
        $this->mail = $mail;
        $this->config = $config;
    }

    /**
     * Function sendEmail
     *
     * Send e-mail depending on if tmp_error.txt is empty or not. (empty = no errors)
     *
     * @return void
     */
    public function sendEmail(): void
    {
        try {
            if (filesize('./log/tmp_error.txt') === 0) {

                $this->setEmailServerSettings();
                $this->mail->isHTML(true);
                $this->mail->Subject = 'BambooHR Connector - Successfully imported';
                $this->mail->Body    = 'Absences were successfully imported!';

                $this->mail->send();
                echo 'Message has been sent';
            } else {
                $this->setEmailServerSettings();
                $this->mail->isHTML(true);
                $this->mail->Subject = 'BambooHR Connector - Error Log';
                $this->mail->Body    = 'An <span style="color: red;">Error</span> has occurred. Please check the
                                        <strong>tmp_error.txt</strong>. <br/>
                                        You can see previous errors inside the <strong>global_error.txt</strong> file';

                $this->mail->addAttachment('./log/tmp_error.txt');
                $this->mail->addAttachment('./log/global_error.txt');

                $this->mail->send();
                echo 'Message has been sent';
            }
        } catch (Exception $e) {
            echo 'Message could not be sent.';
            echo 'Mailer Error: ' . $this->mail->ErrorInfo;
        }
    }

    /**
     * Function setEmailServerSettings
     *
     * Sets the e-mail server settings
     *
     * @return void
     */
    public function setEmailServerSettings(): void
    {
        $this->mail->SMTPDebug = 2;
        $this->mail->isSMTP();
        $this->mail->Host = $this->config->getConfig('mail', 'smtp_host');
        $this->mail->SMTPAuth = $this->config->getConfig('mail', 'smtp_auth');
        $this->mail->Username = $this->config->getConfig('mail', 'username');
        $this->mail->Password = $this->config->getConfig('mail', 'password');
        $this->mail->SMTPSecure = $this->config->getConfig('mail', 'smtp_secure');
        $this->mail->Port = $this->config->getConfig('mail', 'port');

        //Recipients
        $this->mail->setFrom($this->config->getConfig('mail','send_from'), 'Sender');
        $this->mail->addAddress($this->config->getConfig('mail', 'send_to'), 'Recipient');     // Add a recipient
    }
}


Connector.php
<?php

declare(strict_types=1);

namespace BambooHRConnector;

include 'Config.php';
include 'Logger.php';
include 'BambooHRAbsences.php';
include 'BambooHR/API/API.php';
include 'PapershiftAbsences.php';
include 'SendMail.php';

use PHPMailer\PHPMailer\PHPMailer;

require 'vendor/autoload.php';

/**
 * Class Connector
 *
 * @version 2.1
 *
 * @package BambooHRConnector
 * @author Georgi Damyanov <georgi.damyanov@bestit-online.at>
 */
class Connector
{
    /**
     * Instance of PHPMailer
     *
     * @var $mailer PHPMailer
     */
    private $mailer;

    /**
     * Instance of SendMail
     *
     * @var $sendmail SendMail
     */
    private $sendmail;

    /**
     * Instance of Config
     *
     * @var $config Config
     */
    private $config;

    /**
     * Instance of BambooHRAbsences
     *
     * @var $bamboo BambooHRAbsences
     */
    private $bamboo;

    /**
     * Instance of BambooAPI
     *
     * @var $bambooAPI BambooAPI
     */
    private $bambooAPI;

    /**
     * Instance of PapershiftAbsences
     *
     * @var $papershift PapershiftAbsences
     */
    private $papershift;

    /**
     * Instance of Logger
     *
     * @var $logger Logger
     */
    private $logger;

    /**
     * Connector constructor. Create all instances.
     */
    public function __construct()
    {
        $this->logger = new Logger();
        $this->config = new Config($this->logger);
        $this->mailer = new PHPMailer(true);
        $this->sendmail = new SendMail($this->config, $this->mailer);
        $this->bambooAPI = new BambooAPI($this->config->getConfig('bamboohr', 'company'));
        $this->bamboo = new BambooHRAbsences($this->config, $this->bambooAPI, $this->logger, $this->sendmail);
        $this->papershift = new PapershiftAbsences($this->config, $this->logger, $this->sendmail);
    }

    /**
     * Function addAbsencesFromBambooHRToPapershift
     *
     * - Cut/paste error messages from tmp_error.txt to global_error.txt
     * - Get absences from BambooHR
     * - Add the absences to Papershift
     * - Confirm the absences
     * - Send email. (either error or confirmation)
     *
     * @return void
     */
    public function addAbsencesFromBambooHRToPapershift(): void
    {
//        $this->logger->transferDailyLogToGlobal();
        $bambooAbsencesArray = $this->bamboo->getBambooHRAbsences();
        $this->papershift->addAbsencesToPapershift($bambooAbsencesArray);
        $this->papershift->confirmPapershiftAbsences();
        $this->sendmail->sendEmail();
    }
}

$connector = new Connector();
$connector->addAbsencesFromBambooHRToPapershift();


BambooHRAbsences.php
<?php

declare(strict_types=1);

namespace BambooHRConnector;

use Exception;
use SimpleXMLElement;

/**
 * Class BambooHRAbsences
 *
 * @package BambooHRConnector
 * @author Georgi Damyanov <georgi.damyanov@bestit-online.at>
 */
class BambooHRAbsences
{
    /**
     * Instance of SendMail
     *
     * @var $sendmail SendMail
     */
    private $sendmail;

    /**
     * Instance of Logger
     *
     * @var $logger Logger
     */
    private $logger;

    /**
     * The employee data
     *
     * @var $employeesXML SimpleXMLElement
     */
    private $employeesXML;

    /**
     * Instance of BambooAPI
     *
     * @var $bambooAPI BambooAPI
     */
    private $bambooAPI;

    /**
     * Instance of Config
     *
     * @var $config Config
     */
    private $config;

    /**
     * The employee absences
     *
     * @var $absencesArray array
     */
    private $absencesArray = [];

    /**
     * The number of employees
     *
     * @var $employeeCounter int
     */
    private $employeeCounter;

    /**
     * BambooHRAbsences constructor.
     *
     * @param Config $config Config Instance
     * @param BambooAPI $bambooAPI BambooAPI Instance
     * @param Logger $logger Logger Instance
     * @param SendMail $sendmail SendMail Instance
     */
    public function __construct(Config $config, BambooAPI $bambooAPI, Logger $logger, SendMail $sendmail)
    {
        $this->config = $config;
        $this->bambooAPI = $bambooAPI;
        $this->logger = $logger;
        $this->sendmail = $sendmail;
    }

    /**
     * Function getBambooHRAbsences()
     *
     * Fetch the absences from BambooHR
     *  - Go through all the employees
     *  - Get all the absences (with set date) for all the employees
     *
     * @return array Returns an array
     * @throws Exception
     */
    public function getBambooHRAbsences(): array
    {
        $this->bambooAPI->setSecretKey($this->config->getConfig('bamboohr', 'api_token'));

        /** @var BambooHTTPResponse $employeeDirectoryRequestArray */
        $employeeDirectoryRequestArray = $this->bambooAPI->getDirectory();

        $httpcode = $employeeDirectoryRequestArray->statusCode;
        if ($httpcode < 200 || $httpcode > 299) {
            $this->logger->writeErrorToFile('Connection to BambooHR could not be established. Error Code: '.$httpcode);
        }

        $this->employeesXML = $employeeDirectoryRequestArray->getContentXML();
        $this->employeeCounter = $this->getNumberOfEmployees($this->employeesXML);

        while ($this->employeeCounter > -1) {

            if ($this->employeeCounter == 14){
                $employeeExternalID = $this->getEmployeeExternalID($this->employeesXML, $this->employeeCounter);

                /** @var BambooHTTPResponse $timeOffRequestsArray */
                $timeOffRequestsArray = $this->bambooAPI
                    ->getTimeOffRequests(
                        $this->config->getConfig('bamboohr', 'date'),
                        '2019-12-31',
                        $this->config->getConfig('bamboohr', 'status'), '',
                        $employeeExternalID
                    );
                $timeOffRequestsXML = $timeOffRequestsArray->getContentXML();

                var_dump($timeOffRequestsXML);exit;

                for ($i = 0; $i < sizeof($timeOffRequestsXML->request); $i++) {
                    $absence = [
                        'id' => $employeeExternalID,
                        'name' => (string) $timeOffRequestsXML->request[$i]->employee,
                        'start' => (string) $timeOffRequestsXML->request[$i]->start,
                        'end' => (string) $timeOffRequestsXML->request[$i]->end,
                        'type' => (string) $timeOffRequestsXML->request[$i]->type,
                        'unit' => (string) $timeOffRequestsXML->request[$i]->amount->attributes()['unit'],
                        'amount' => (string) $timeOffRequestsXML->request[$i]->amount,
                        'full_day' => false
                    ];

                    if ($this->config->getConfig('bamboohr', 'filter_enabled') === false ||
                        !in_array($absence['type'], $this->config->getConfig('bamboohr', 'filter_array'))
                    ) {
                        $this->absencesArray[] = $absence;
                    }
                }
            }
            $this->employeeCounter--;
        }

        var_dump($this->absencesArray);

        exit;

        return $this->absencesArray;
    }

    /**
     * Function getNumberOfEmployees
     *
     * Get the number of employees
     *
     * @param SimpleXMLElement $simplexml Array with the employees
     *
     * @return int returns (number of employees - 1) (because count starts from 0)
     */
    private function getNumberOfEmployees($simplexml): int
    {
        return count($simplexml->employees->employee)-1;
    }

    /**
     * Function getEmployeeExternalID
     *
     * Get the external ID of an employee
     *
     * @param SimpleXMLElement $simplexml Array with the employees
     * @param int $index employee counter
     *
     * @return int returns the externalID of an employee
     */
    private function getEmployeeExternalID($simplexml, $index): int
    {
        return (int) $simplexml->employees->employee[$index]->attributes()['id'];
    }
}


PapershiftAbsences.php
<?php

declare(strict_types=1);

namespace BambooHRConnector;

use Exception;

/**
 * Class PapershiftAbsences
 *
 * @package BambooHRConnector
 * @author Georgi Damyanov <georgi.damyanov@bestit-online.at>
 */
class PapershiftAbsences
{
    /**
     * Instance of SendMail
     *
     * @var $sendmail SendMail
     */
    private $sendmail;

    /**
     * Instance of Logger
     *
     * @var $logger Logger
     */
    private $logger;

    /**
     * Instance of Config
     *
     * @var $config Config
     */
    private $config;

    /**
     * PapershiftAbsences constructor.
     *
     * @param Config $config Config Instance
     * @param Logger $logger Logger Instance
     * @param SendMail $sendmail SendMail Instance
     */
    public function __construct(Config $config, Logger $logger, SendMail $sendmail)
    {
        $this->config = $config;
        $this->logger = $logger;
        $this->sendmail = $sendmail;
    }

    /**
     * Function addAbsencesToPapershift
     *
     * - Go through all the absences
     * - Convert them to json
     * - Add them to Papershift
     *
     * @param array $array Contains all absences from Bamboo
     *
     * @return void
     * @throws Exception
     */
    public function addAbsencesToPapershift($array): void
    {
        $ch = curl_init($this->config->getConfig('papershift', 'api_absences_url'));
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'POST');

        for ($i = 0; $i < sizeof($array); $i++) {
            $convertedArrayElem = $this->convertTimeToISO8601($array[$i]);
            $papershiftAbsence = $this->createPapershiftAbsencesArray($convertedArrayElem);

            $JSONpapershiftAbsence = $this->fromArrayToJson($papershiftAbsence);

            curl_setopt($ch, CURLOPT_POSTFIELDS, $JSONpapershiftAbsence);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_HTTPHEADER, [
                'Content-Type: application/json',
                'Content-Length: ' . strlen($JSONpapershiftAbsence),
                'Accept: application/json'
            ]);
            curl_setopt($ch, CURLOPT_TIMEOUT, 5);
            curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);

            $result = curl_exec($ch);
            $response = curl_getinfo($ch, CURLINFO_HTTP_CODE);

            if ($response < 200 || $response > 299) {
                $this->logger->writeErrorToFile('Result: '.$result.' | Error Code: '.$response);
            }
        }
    }

    /**
     * Function confirmPapershiftAbsences()
     *
     * - Call getPapershiftAbsences() to get all absences for the selected date
     * - Get the ID's for all absences
     * - Confirm
     *
     * @return void
     * @throws Exception
     */
    public function confirmPapershiftAbsences(): void
    {
        $absencesArray = $this->getPapershiftAbsences();
        $ch = curl_init($this->config->getConfig('papershift', 'api_absences_confirm_url'));

        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');

        for($i = 0; $i < sizeof($absencesArray['absences']); $i++) {
            $absenceJSON = $this->fromArrayToJson([
                'api_token' => $this->config->getConfig('papershift', 'api_token'),
                'absence' => [
                    'id' => $absencesArray['absences'][$i]['id']
                ]
            ]);

            curl_setopt($ch, CURLOPT_POSTFIELDS, $absenceJSON);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_HTTPHEADER, [
                'Content-Type: application/json',
                'Content-Length: ' . strlen($absenceJSON),
                'Accept: application/json'
            ]);
            curl_setopt($ch, CURLOPT_TIMEOUT, 5);
            curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);

            $result = curl_exec($ch);
            $response = curl_getinfo($ch, CURLINFO_HTTP_CODE);

            if ($response < 200 || $response > 299) {
                $this->logger->writeErrorToFile('Result: '.$result.' | Error Code: '.$response);
            }
        }
    }

    /**
     * Function getPapershiftAbsences()
     *
     * Get all absences between range_start and range_end (ISO) and return them as an array
     *
     * @return array
     */
    public function getPapershiftAbsences(): array
    {
        $ch = curl_init($this->config->getConfig('papershift', 'api_absences_url'));
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'GET');

        $requestAbsence = $this->fromArrayToJson([
            'api_token' => $this->config->getConfig('papershift','api_token'),
            'range_start' => $this->config->getConfig('bamboohr', 'date'),
            'range_end' => $this->config->getConfig('bamboohr', 'date')
        ]);

        curl_setopt($ch, CURLOPT_POSTFIELDS, $requestAbsence);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_HTTPHEADER, [
            'Content-Type: application/json',
            'Content-Length: ' . strlen($requestAbsence),
            'Accept: application/json'
        ]);
        curl_setopt($ch, CURLOPT_TIMEOUT, 5);
        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);

        $result = curl_exec($ch);
        $resultArray = $this->fromJsonToArray($result);

        return $resultArray;
    }

    /**
     * Function convertTimeToISO8601
     *
     * Convert absence time from Y-m-d to ISO8601 (required by Papershift)
     *
     * The function gmdate converts hours to h:i:s (f.e 2.5556h = 02:33:20)
     *
     * @param array $array Takes the Bamboo-Absences array
     *
     * @return array Returns same array with ISO8601 time
     */
    public function convertTimeToISO8601($array): array
    {
        if ($array['unit'] === 'days') {
            $array['start'] = (string) $array['start'].$this->config->getConfig('papershift', 'start_time_days');
            $array['end'] = (string) $array['end'].$this->config->getConfig('papershift', 'end_time_days');
            $array['full_day'] = true;
        } elseif ($array['unit'] === 'hours') {
            $array['start'] = (string) $array['start'].$this->config->getConfig('papershift', 'start_time_hours');
            $array['end'] = (string) $array['end'].'T'.gmdate('H:i:s', (int) floor($array['amount'] * 3600)).'+01:00';
        }
        return $array;
    }

    /**
     * Function createPapershiftAbsencesArray
     *
     * Convert BambooHR array to Papershift array with the correct structure
     *
     * @param array $array Takes the Bamboo-Absences array
     *
     * @return array Returns a valid structured Papershift array
     */
    public function createPapershiftAbsencesArray($array): array
    {
        return [
            'api_token' => $this->config->getConfig('papershift', 'api_token'),
            'absence' => [
                'absence_type_external_id' => $array['type'],
                'user_external_id' => (string) $array['id'],
                'starts_at' => (string) $array['start'],
                'ends_at' => $array['end'],
                'full_day' => $array['full_day']
            ]
        ];
    }

    /**
     * Function fromArrayToJson
     *
     * Convert an array to json
     *
     * @param array $array Takes an array
     *
     * @return string Returns a json string
     */
    public function fromArrayToJson($array): string
    {
        return json_encode($array, JSON_UNESCAPED_UNICODE);
    }

    /**
     * Function fromJsonToArray()
     *
     * Convert a JSON string to an array
     *
     * @param string $json Takes a JSON string
     *
     * @return array Returns an array
     */
    public function fromJsonToArray($json): array
    {
        return json_decode($json, true);
    }
}
